### ISSUE TITLE
Text component for Canvas element doesn't redraw correctly without a width or height

### ISSUE THREAD

**Issue by kcaze** [craftyjs/Crafty#498]
When I have an entity with a Sprite component, then calling text() fails to update any Canvas entities with Text components (DOM seems to work fine).

Here's a fiddle illustrating the problem: http://jsfiddle.net/kcaze/x8t2B/1/

**Comment by starwed** [comment]
Ah, I think the issue is actually that the text entity isn't redrawn correctly without a height or width.

When it is the only element, a full redraw is done, so the text changes correctly.  When you add another canvas  element (it doesn't matter what type!) a partial redraw is done instead.  Because the text entity lacks dimensions, it is skipped during this process -- and that's the bug.

I'll work on a fix for that, but a workaround in the meantime might be to give the canvas text a width/height.

**Comment by kcaze** [comment]
As a workaround, in the text() function, I obtain the canvas's context with Crafty.canvas.context and then set the width using that. The height is simply the font size.

Another thing I've noticed is that the draw() method for the text draws the text at (obj._x, obj._y + obj._h). I don't quite understand why the text would be drawn at the y position plus the height, rather than simply at the y position. It seems like the clearing of the dirty rectangles in the DrawManager doesn't clear it properly if the text is drawn at (obj._x, obj._y) though.

EDIT: I realized that adding the height was necessary in order to compensate for the fact that the textBaseline is the bottom of the text by default and not the top. In my pull request, I've fixed that, and had text() update the width and height each time it's called.
